File: leetcode/tree/bst/0098_validate_bst.py
# LC 98. Validate Binary Search Tree | Medium
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val; self.left = left; self.right = right

class Solution:
    def isValidBST(self, root, lo=float('-inf'), hi=float('inf')):
        if not root: return True
        if root.val <= lo or root.val >= hi: return False
        return self.isValidBST(root.left, lo, root.val) and self.isValidBST(root.right, root.val, hi)

File: leetcode/tree/bst/0098_validate_bst.cpp
// LC 98. Validate Binary Search Tree | Medium
#include <climits>
struct TreeNode { int val; TreeNode *left, *right; TreeNode(int x) : val(x), left(nullptr), right(nullptr) {} };

class Solution {
    bool check(TreeNode* n, long lo, long hi) {
        if (!n) return true;
        if (n->val <= lo || n->val >= hi) return false;
        return check(n->left, lo, n->val) && check(n->right, n->val, hi);
    }
public:
    bool isValidBST(TreeNode* root) { return check(root, LONG_MIN, LONG_MAX); }
};
